lib/ok.ex

```diff
@@ -83,6 +83,38 @@ defmodule OK do
     if is_list(result), do: {:ok, Enum.reverse(result)}, else: result
   end
 
+  @doc """
+  Takes a result tuple, a predicate function, and an optional error reason.
+  If the result tuple is tagged as a success then its value will be passed to the predicate function.
+  If the predicate returns `true`, then the result tuple stay the same.
+  If the predicate returns `false, then the result tuple becomes `{:error, reason}`.
+  If the tag is failure then the predicate function is skipped.
+
+  ## Examples
+
+      iex> OK.filter({:ok, 2}, fn (x) -> x == 2 end)
+      {:ok, 2}
+
+      iex> OK.filter({:ok, 2}, fn (x) -> x == 3 end, :bad_value)
+      {:error, :bad_value}
+
+      iex> OK.filter({:error, :some_reason}, fn (x) -> x == 4 end)
+      {:error, :some_reason}
+  """
+  @spec filter({:ok, a} | {:error, reason}, (a -> boolean), test_failure_reason) ::
+          {:ok, a} | {:error, test_failure_reason} | {:error, reason}
+        when a: any, reason: any, test_failure_reason: any
+  def filter(result_tuple, func, reason \\ :failed_predicate)
```

Comment:
Thinking some more about this I think it will be really hard to debug if you start getting a `{error, failed_predicate}` from somewhere you don't know. I think it would be more helpful to require a reason. i.e. not set the default.
In which case we don't need the test I mentioned above